From f0eabb60886e9adf8e8a3d1201eb3b9b6e49de57 Mon Sep 17 00:00:00 2001
From: Jiasheng Jiang <jiashengjiangcool@outlook.com>
Date: Tue, 6 Aug 2024 19:18:34 +0000
Subject: [PATCH 110/125] test/provider_test.c: Add OSSL_PROVIDER_unload() to
 avoid memory leak

Add OSSL_PROVIDER_unload() when OSSL_PROVIDER_add_builtin() fails to avoid memory leak.

Fixes: 5442611dff ("Add a test for OSSL_LIB_CTX_new_child()")
Signed-off-by: Jiasheng Jiang <jiashengjiangcool@outlook.com>

Reviewed-by: Shane Lontis <shane.lontis@oracle.com>
Reviewed-by: Tomas Mraz <tomas@openssl.org>
(Merged from https://github.com/openssl/openssl/pull/25109)

(cherry picked from commit 55662b674543c9385600bc9b7c46277ef69b4dba)
---
 test/provider_test.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/test/provider_test.c b/test/provider_test.c
index 2d20d12071..3dc5a2109e 100644
--- a/test/provider_test.c
+++ b/test/provider_test.c
@@ -255,6 +255,7 @@ static int test_builtin_provider_with_child(void)
 
     if (!TEST_true(OSSL_PROVIDER_add_builtin(libctx, name,
                                              PROVIDER_INIT_FUNCTION_NAME))) {
+        OSSL_PROVIDER_unload(legacy);
         OSSL_LIB_CTX_free(libctx);
         return 0;
     }
-- 
2.39.5 (Apple Git-154)

